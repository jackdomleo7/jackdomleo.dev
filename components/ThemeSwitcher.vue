<template>
  <fieldset>
    <legend>Switch theme</legend>

    <div class="theme-options">
      <div class="theme-option">
        <div>
          <input v-model="$colorMode.preference" :id="`${id}-theme-system`" type="radio" :name="`${id}-theme`" value="system" />
          <svg-icon name="monitor" />
        </div>
        <label :for="`${id}-theme-system`">System</label>
      </div>

      <div class="theme-option">
        <div>
          <input v-model="$colorMode.preference" :id="`${id}-theme-light`" type="radio" :name="`${id}-theme`" value="light" />
          <svg-icon name="sun" />
        </div>
        <label :for="`${id}-theme-light`">Light</label>
      </div>

      <div class="theme-option">
        <div>
          <input v-model="$colorMode.preference" :id="`${id}-theme-dark`" type="radio" :name="`${id}-theme`" value="dark" />
          <svg-icon name="moon" />
        </div>
        <label :for="`${id}-theme-dark`">Dark</label>
      </div>
    </div>
  </fieldset>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  name: 'ThemeSwitcher',
  props: {
    id: {
      type: String,
      required: true
    }
  }
})
</script>

<style lang="scss" scoped>
fieldset {
  border: none;
}

legend {
  font-size: var(--text-body);
  margin-bottom: 0.5rem;
}

.theme-options {
  display: flex;
  gap: 1rem;
}

.theme-option {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.25rem;
  font-size: var(--text-small);

  > div {
    display: inline-flex;
    flex-direction: column;
    align-items: center;
    position: relative;
  }

  svg {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 60%;
    color: var(--colour-text-secondary);
  }

  input {
    appearance: none;
    display: grid;
    place-items: center;
    height: 3rem;
    width: 3rem;
    border: 2px solid var(--colour-text-secondary);
    border-radius: 0.625rem;
    cursor: pointer;

    &:checked {
      border-color: var(--colour-primary);
      background-color: var(--colour-primary-faint);

      + svg {
        color: var(--colour-primary);
      }
    }
  }
}
</style>
