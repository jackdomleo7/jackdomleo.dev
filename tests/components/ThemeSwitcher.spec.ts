import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import ThemeSwitcher from '@/components/ThemeSwitcher.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('ThemeSwitcher', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(ThemeSwitcher, {
      mocks: {
        $colorMode: {
          preference: 'system'
        }
      },
      propsData: {
        id: 'mobile-theme-switcher'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(ThemeSwitcher, {
      mocks: {
        $colorMode: {
          preference: 'system'
        }
      },
      propsData: {
        id: 'mobile-theme-switcher'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(ThemeSwitcher, {
      mocks: {
        $colorMode: {
          preference: 'system'
        }
      },
      propsData: {
        id: 'mobile-theme-switcher'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(ThemeSwitcher, {
      mocks: {
        $colorMode: {
          preference: 'system'
        }
      },
      propsData: {
        id: 'mobile-theme-switcher'
      },
      stubs: ['svg-icon']
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Props
   */

  it('renders the id', () => {
    const wrapper = mount(ThemeSwitcher, {
      mocks: {
        $colorMode: {
          preference: 'system'
        }
      },
      propsData: {
        id: 'mobile-theme-switcher'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.find('input#mobile-theme-switcher-theme-system[name="mobile-theme-switcher-theme"]').exists()).toBe(true)
    expect(wrapper.find('label[for="mobile-theme-switcher-theme-system"]').exists()).toBe(true)
    expect(wrapper.find('input#mobile-theme-switcher-theme-light[name="mobile-theme-switcher-theme"]').exists()).toBe(true)
    expect(wrapper.find('label[for="mobile-theme-switcher-theme-light"]').exists()).toBe(true)
    expect(wrapper.find('input#mobile-theme-switcher-theme-dark[name="mobile-theme-switcher-theme"]').exists()).toBe(true)
    expect(wrapper.find('label[for="mobile-theme-switcher-theme-dark"]').exists()).toBe(true)
  })
})
