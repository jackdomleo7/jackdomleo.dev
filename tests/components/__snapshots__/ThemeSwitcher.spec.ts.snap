// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher matches snapshot when mounted 1`] = `
<fieldset>
  <legend>
    Switch theme
  </legend>
   
  <div
    class="theme-options"
  >
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-system"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="system"
        />
         
        <svg-icon-stub
          name="monitor"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-system"
      >
        System
      </label>
    </div>
     
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-light"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="light"
        />
         
        <svg-icon-stub
          name="sun"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-light"
      >
        Light
      </label>
    </div>
     
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-dark"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="dark"
        />
         
        <svg-icon-stub
          name="moon"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-dark"
      >
        Dark
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`ThemeSwitcher matches snapshot when shallow mounted 1`] = `
<fieldset>
  <legend>
    Switch theme
  </legend>
   
  <div
    class="theme-options"
  >
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-system"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="system"
        />
         
        <svg-icon-stub
          name="monitor"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-system"
      >
        System
      </label>
    </div>
     
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-light"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="light"
        />
         
        <svg-icon-stub
          name="sun"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-light"
      >
        Light
      </label>
    </div>
     
    <div
      class="theme-option"
    >
      <div>
        <input
          id="mobile-theme-switcher-theme-dark"
          name="mobile-theme-switcher-theme"
          type="radio"
          value="dark"
        />
         
        <svg-icon-stub
          name="moon"
        />
      </div>
       
      <label
        for="mobile-theme-switcher-theme-dark"
      >
        Dark
      </label>
    </div>
  </div>
</fieldset>
`;
